WCSAxes AxisLabel refactoring

dstansby opened the pull request:
As I dive deep into the rabbit hole that is https://github.com/astropy/astropy/pull/10797, I have refactored some code along the way. This is probably easier to review one commit at a time

- Move the code to calculate the distance halfway along a spine to the spine class itself.
- Move code outside a loop that doesn't use the loop variable
- De-duplicate a check in an if statement that occurs 4 times
- Cleanup the logic used in calculating the position of tick labels

Cadair commented on astropy/visualization/wcsaxes/axislabels.py:
```diff
@@ -105,37 +105,30 @@ def draw(self, renderer, bboxes, ticklabels_bbox,
                     coord_ticklabels_bbox[axis] = [None]
 
                 visible = axis in visible_ticks and coord_ticklabels_bbox[axis][0] is not None
+                x = xcen
+                y = ycen
```
Why redfine these here? why not just call them `x` & `y` from the outset?

Cadair reviewed (approved):
Makes sense to me.

pllim commented:
I am going to close/re-open to kick off RTD again to make sure the timeout isn't related.

pllim commented:
@Cadair , if you are not waiting for someone else to review, feel free to merge when CI passes.

Cadair commented:
hmmm, it timed out again.

dstansby commented:
Rebased to see if the tests will pass; I also dropped the first commit, since on a second viewing it looks like that might not be a valid change to make (even though the tests passed).

Cadair commented:
@dstansby should this be a bugfix as well? is it needed for the bbox quest?

dstansby commented:
I think the bbox quest is kind of a bugfix, but in reality will require so much refactoring (and is a beaviour change, albeit from an incorrect to a correct behaviour) it's probably best to put it in a major release?
